cloudevents/working-drafts/minor-versions.md

```diff
@@ -0,0 +1,337 @@
+an extension attribute as the consumer is aware of all core v1.0 attribtes,
+and `xyz` appearing in a v1.1 event which is *either* an extension attribute *or* a new optional
+core attribute.)
+
+## SDKs
+
+If we adopt minor versions, SDKs should change to support them, so that (for example) an SDK
+supporting v1.1 will *also* parse v1.2 events. We should keep track of which SDKs
+(and which versions of those SDKs) support parsing v1.x events.
+
+Different SDKs expose features very differently. This document is written from the perspective
+of the C# SDK, but other viewpoints would be highly valued: please add feedback, particularly
+if the approach suggested doesn't make sense for an SDK you're familiar with.
+
+### Creating events: default version handling
+
+It may sound natural to allow something like:
+
+```csharp
+var evt = new CloudEvent { Id = "..." };
+```
+
+... to just use the latest spec version that the SDK supports. However, this is a breaking change.
+Consider:
+
+```csharp
+var evt = new CloudEvent
+{
+    Id = "...",
+    ["xyz"] = "test"
+};
+```
+
+When the SDK only supports v1.0, that creates a valid event. When the SDK supports v1.1 as well,
+that would produce an invalid event or throw an exception, if the version is implicitly v1.1.
+Instead, the SDK should *implicitly* use the latest version that was present when the first stable
+version of the SDK came out, within its own current major version. In other words, if v1.1 ships
+and a new SDK is created, that can default to v1.1 - and if an SDK creates its own new major version,
+*that* can default to v1.1 as well... but in both cases that default should stay as v1.1 in new minor
+SDK versions even after v1.2 ships.
+
+An SDK *may* choose to provide a way of explicitly opting into the "latest supported" version:
+
+```csharp
+var evt = new CloudEvent(CloudEventsSpecVersion.Latest)
+{
+  
+};
+```
+
+### Accessing attributes
+
+If an SDK exposes core attributes in a distinct way from extension attributes (as shown above,
+where `Id` is a strongly-typed property, whereas `["xyz"]` uses an indexer with a string literal)
+then any attempt to use the "special" access should fail when accessing an attribute within
+an event with a version that doesn't have that attribute as a core attribute. Example:
+
+```csharp
+// Create a v1.0 event
+var evt = new CloudEvent(CloudEventsSpecVersion.V1_0);
+
+// Try to access the new Xyz property, referring to the optional attribute introduced in
+// v1.1: this should fail.
+evt.Xyz = 1;
+
+// Access the xyz attribute by name, as an extension attribute.
+// This should work, even though the value wouldn't be valid in a v1.1 event.
+evt["xyz"] = "test";
+
+// Likewise accessing the property for an event parsed from elsewhere:
+evt = httpRequest.ParseCloudEvent();
+// If evt is v1.0, this should fail *even if it has an xyz attribute value that's valid in v1.1.
+int xyz = evt.Xyz;
```

Comment:
again, I disagree if the SDK chooses to know about it as a first class attribute.
Yes they're taking a risk the same way the extension author was taking a risk by using a name that the spec might use later on... but that's a risk they're willing to take. And that risk is always there w.r.t. conflicting extensions.

Reply:
Yes the SDK user is the one who feels the pain, but "taking on the risk" is an interesting statement because you seem to be assuming that they know the boundaries of CE 1.0 and therefore knows that "xyz" isn't part of CE 1.0 - they might not. To them, CE core might be whatever the SDK exposes as a core property. Much in the same way a compiler user may not bother to check to see if a certain features of their compiler/language is part of that language's spec or it's an extension that this particular compiler added.

But either way... yes there's risk/pain associated with all of this and it's true no matter what we do - even if we never create a new 1.x optional attribute.

Reply:
I would agree this would fail for a 1.0 event. The type `evt` is not aware of `Xyz` in the implementation, so the code should not compile in a strongly typed language.

Reply:
See comment above about extensions - not all CE processing code will come from us.

The mental model I have is similar to how some C compilers add extensions to the ISO standard C definition. Ignoring flags like a --strict type of compilation mode, the extensions appear on-par with the standard stuff. This feels like the same type of thing to me.